### PR TITLE
fix MDC scope nesting

### LINKED ISSUE #8722: MDC is empty after a Jaeger sub span is closed.
**Describe the bug**
After a subspan closes MDC is empty.

**Expected behavior**
MDC context should contain traceId, spanId and sample of parent span.

**Actual behavior**
The MDC context is wiped on MDCScope.close() without restoring the MDC info the parent span.

**To Reproduce**
1. Create Trace A 
2. Within Trace A create Trace B
3. Within Trace A inspect MDC after Trace B closes
4. MDC is empty.

### PR DESCRIPTION
**Pull request by Ladicek**
Fixes #8722.

### CODE REVIEW

**Comment by Ladicek** [comment]
Draft to let CI run in my fork.

**Comment by gsmet** [comment]
@pavolloffay any chance you could have a look at this PR?

**Comment by pavolloffay** [comment]
Sure I will have a look.

**Comment by Ladicek** [comment]
CI passed in my fork (baring 1 failure which seems totally unrelated to me), un-WIP-ing.

**Comment by gsmet** [comment]
@pavolloffay just to be sure you have the timeline in mind, I would need to merge that before end of day to get this into 1.4.

File: extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/NestedMdcScopesTest.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import org.junit.jupiter.api.Test;
+
+import io.jaegertracing.internal.JaegerSpanContext;
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.util.ThreadLocalScopeManager;
+import io.quarkus.jaeger.runtime.MDCScopeManager;
+
+public class NestedMdcScopesTest {
```
Comment (pavolloffay) [inline extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/NestedMdcScopesTest.java:18]: Could be this test moved to Jaeger artifact - close to the MDC implementation?

Comment (Ladicek) [reply extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/NestedMdcScopesTest.java:18]: Hmm, I'm not sure why I added it to `smallrye-opentracing`, indeed this should be in `jaeger`. Thanks :-)

Comment (Ladicek) [reply extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/NestedMdcScopesTest.java]: Done.

File: extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/MDCScope.java [hunk 2]
```diff
@@ -33,6 +35,10 @@ public void close() {
         MDC.remove(TRACE_ID);
         MDC.remove(SPAN_ID);
         MDC.remove(SAMPLED);
+
+        if (toRestore != null && toRestore.span().context() instanceof JaegerSpanContext) {
```
Comment (pavolloffay) [inline extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/MDCScope.java:39]: This certainly fixes the problem, but I am wondering whether it isn't better to just restore the old values of the keys that were removed. For instance they could be explicitly set by a user or other integration.

**Review by pavolloffay — APPROVED** [review]

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I will merge that one and backport to 1.4.

Further improvements can be made in 1.5.

File: extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/MDCScope.java (hunk 2 above)
Comment (Ladicek) [reply extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/MDCScope.java:39]: Hmm, yea, we could do that...

Comment (Ladicek) [reply extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/MDCScope.java:39]: I've followed the `ThreadLocalScopeManager` / `ThreadLocalScope` example :-)

But I agree it's cleaner to just restore the MDC map content. I have also found that the test can be improved a bit. I will submit a subsequent PR.

**Pull request merged**